[autochunker] support gradient accumulation

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #171477
* __->__ #171359

With gradient accumulation, there will be a fx.Node dividing the loss by a scalar (the gradient accumulation steps). Update the propagation rule to handle that.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

shunting314 committed:
[autochunker] support gradient accumulation

[ghstack-poisoned]

eellison commented on test/inductor/test_auto_chunker.py:
```diff
@@ -176,6 +183,74 @@ def f(x, y):
+
+        self.assertTrue(same(expect, actual, tol=1e-3), f"{expect=}\n{actual=}")
+
+        if config.auto_chunker.enable:
```
shouldnt this always be enabled in this test?

shunting314 replied:
I occasionally run this test with AutoChunker disabled leveraging this decorator in this test class:
```
@config.patch(
    "auto_chunker.enable", os.environ.get("TORCHINDUCTOR_AUTO_CHUNKER", "1") == "1"
)
```

That's mainly used in early development to check compile behavior without chunking though.

shunting314 committed:
Update on "[autochunker] support gradient accumulation"


With gradient accumulation, there will be a fx.Node dividing the loss by a scalar (the gradient accumulation steps). Update the propagation rule to handle that.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]
